lib/puppet/configurer.rb

```diff
@@ -126,6 +126,82 @@ def convert_catalog(result, duration, facts, options = {})
     catalog
   end
 
+  def check_fact_name_length(name)
+    unless Puppet[:max_factnamelength] == -1 
+      if Puppet[:max_factnamelength] == 0
+        Puppet[:max_factnamelength] = 100
+        if name.length > Puppet[:max_factnamelength]
+          Puppet.warning _("Fact %{name} exceedes the lenght limit: %{limit}") % { name: name, limit: Puppet[:max_factnamelength] }
+        end
+      else
+        if name.length > Puppet[:max_factnamelength]
+          Puppet.warning _("Fact %{name} exceedes the lenght limit: %{limit}") % { name: name, limit: Puppet[:max_factnamelength] }
+        end 
+      end
+    end
+  end
+
+  def check_fact_values_length(keys,values)
+    unless Puppet[:max_factvaluelength] == -1 
+      if Puppet[:max_factvaluelength] == 0
+        Puppet[:max_factvaluelength] = 150
+        if values.to_s.length > Puppet[:max_factvaluelength]
+          Puppet.warning _("Fact '%{k}' with value '%{v}' exceeds the value length limit: %{limit}") % { k: keys, v: values, limit: Puppet[:max_factvaluelength] } 
+        end
+      else
+        if values.to_s.length > Puppet[:max_factvaluelength]
+          Puppet.warning _("Fact '%{k}' with value '%{v}' exceeds the value length limit: %{limit}") % { k: keys, v: values, limit: Puppet[:max_factvaluelength] }
+        end 
+      end
+    end
+  end
+
+  def get_fact_values(keys, values)
+    if values.is_a?(Hash) 
+      values.each do |k,v|
+        value = v || k
+        if value.is_a?(Hash) || value.is_a?(Array)
+          get_fact_values(keys, value)
+        else
+          check_fact_values_length(k, v)
+        end
+      end
+    elsif values.is_a?(Array)
+      values.each do |value|
+        if value.is_a?(Hash) || value.is_a?(Array)
+          get_fact_values(keys, value)
+        else
+          check_fact_values_length(keys, value)
+        end
+      end
+    else
+      check_fact_values_length(keys, values)
+    end
+  end
+
+  def check_size_limits(size)
+    unless Puppet[:max_toplevelfacts] == -1 
+      if Puppet[:max_toplevelfacts] == 0
+        Puppet[:max_toplevelfacts] = 200
+        if size > Puppet[:max_toplevelfacts]
+          Puppet.warning _("You have exceeded the number of top facts limit: %{limit}") % { limit: Puppet[:max_toplevelfacts] }
+        end
+      else
+        if size > Puppet[:max_toplevelfacts]
+          Puppet.warning _("You have exceeded the number of top facts limit: %{limit}") % { limit: Puppet[:max_toplevelfacts] }
+        end 
+      end
+    end
+  end
+
+  def check_top_level(facts)
+    check_size_limits(facts.size)
+    for key, value in facts do
```

Comment:
Do we care about nested facts keys? for example this will pass the fact name validation:
```
 facts.values = {'1' => { 'my_super_long_name_with_custom_value' => 'my_new_fact_value'} }
```
because the name is validated only on the first level